### ISSUE TITLE
Converting pretrained Nemo T5/UL2 to Huggingface (request for help)

### ISSUE THREAD

**Issue by Lauler** [NVIDIA/NeMo#6626]
**Is your feature request related to a problem? Please describe.**

An important consideration for us when deciding which library, framework or model implementation to use when pretraining models is whether the model will be accessible and easy to use for end users. In practice this translates to whether it is possible and straightforward to convert the model to a Huggingface compatible format. 

For Nemo, there exists a conversion script to [convert a model from Huggingface to Nemo format](https://github.com/NVIDIA/NeMo/blob/main/scripts/nlp_language_modeling/hf_t5-v1_1_to_nemo.py). However, there doesn't exist a resource for a conversion in the opposite direction. 

Since an `HF -> Nemo` script existed, we thought converting from `Nemo -> HF` would simply be a matter of reversing that script. We therefore [carefully configured our Nemo configs](https://huggingface.co/Lauler/UL2-nemo-conversion/tree/main/nemo_config/ul2-base-nl36) to return a model that weight wise seemed compatible with Huggingface's T5 implementation (e.g. no bias terms, tied encoder/decoder word embeddings). Then we wrote our [own conversion script](https://huggingface.co/Lauler/UL2-nemo-conversion/blob/main/convert_nemo_ul2_checkpoint.py) to reverse Nemo's existing one. 

Our script appears to work fine reversing the weight conversion performed in Nemo's `HF -> Nemo` script. Unfortunately, it does not work as well when attempting to convert a model initialized and pretrained with Nemo Megatron. 

We are aware that [Megatron-LM uses different ordering of QKV](https://github.com/NVIDIA/Megatron-LM/blob/42c1cf4279acea5a554500dcb552211f44cbec45/megatron/checkpointing.py#L209-L237) in the attention layers depending on the version of Megatron-LM used. We are also aware of an existing conversion script that Huggingface have created for converting Megatron-BERT to Huggingface, where they adapt the ordering of QKV in Megatron to [match the ordering used in Huggingface](https://github.com/NVIDIA/Megatron-LM/blob/42c1cf4279acea5a554500dcb552211f44cbec45/megatron/checkpointing.py#L209-L237). We copied this reordering, and added a `--fix_qkv` argument in our conversion script that applies the same reordering of QKV as Huggingface does. 

Unfortunately, none of the above solves the issue we have with our conversion script.

**Describe the solution you'd like**

We've uploaded Nemo model checkpoints, conversion scripts, and detailed instructions in the README on how to reproduce our attempted conversion in [this Huggingface repo](https://huggingface.co/Lauler/UL2-nemo-conversion
). We'd very much appreciate some help figuring out how T5 and UL2 models can be converted from Nemo Megatron to HF. Specifically:

1. Is our Nemo config appropriate to be able to perform a weight conversion? 
2. Does the issue lie with the QKV reordering in Megatron? What is the correct way of reordering QKV in self-attention and KV in cross attention for encoder-decoders? How does one find out which Megatron "checkpoint version" one's saved checkpoints are using?

To clone the HF repo:

```bash
git clone https://huggingface.co/Lauler/UL2-nemo-conversion
```

**Describe alternatives you've considered**

We searched for [UL2](https://huggingface.co/models?search=ul2) and T5 models on Huggingface to see whether we could find any that were successfully converted from Megatron-LM or Nemo. All the HF compatible ones we could find were either FLAX/JAX or models with unclear origin. Nvidia's own uploaded T5 models consist of [`.nemo` checkpoints with instructions to install Apex and Nemo](https://huggingface.co/nvidia/nemo-megatron-t5-3B).

Ideally, we'd like to use Nemo to pretrain language models. But realistically we are going to go for the option that means our users are able to use the final model. Nemo is a great library, but the reality is only a minority of our users would be able to use the models without the kind of well documented convenience functions that exist in HF.

So in other words: our main alternative at this point is switching to JAX/FLAX.  

**Additional context**

Would it be more appropriate to open an issue about this in the Megatron-LM repo rather than here?

**Comment by MaximumEntropy** [comment]
Hey! Apologies, I haven't gone through your full repository in detail, but I had a few questions and potential differences I see (besides what you've already mentioned):

1. Once you convert from NeMo -> HF, are you trying to run inference with the `T5ForConditionalGeneration` class in HF? If yes, there are several other differences between NeMo's T5 implementation and HF/Google.
2. HF/Google fold the division by sqrt(hidden_dim) of QK^T into the weight initialization so instead of doing QK^T/sqrt(hidden_dim), they do just QK^T. In NeMo, we still do the division by sqrt and simply have a flag `normalize_attention_scores=False` to optionally skip it when working with models converted from HF.
3. NeMo uses LayerNorm by default instead of RMSNorm in HF/Google.
4. Make sure the RPE configs are correct.
5. Make sure you are using the geglu activation function.

Have you tried "cyclic" conversion (i.e.) start from HF -> NeMo -> HF and verify that you are seeing identical outputs from the first and last HF checkpoints?

**Comment by Lauler** [comment]
Hi! Thanks for the feedback on things to look out for. I'll try answering your questions

1. After I convert the model NeMo -> HF, I do indeed load the model via T5ForConditionalGeneration and compare against NeMo inference. See my [inference script here](https://huggingface.co/Lauler/UL2-nemo-conversion/blob/main/test_ul2_hf.py). 
2. We trained with `normalize_attention-scores=True`. I did read the comment in the NeMo config file regarding this setting. However, it was a bit unclear whether the comment referred only to HF models converted to NeMo for which a user wanted to continue pretraining, or if it was equally necessary for a model trained ground up via NeMo to later be converted to HF. Will try setting this to `False`!
3. We changed to RMSNorm in [our encoder/decoder NeMo config](https://huggingface.co/Lauler/UL2-nemo-conversion/blob/main/nemo_config/ul2-base-nl36/megatron_model_ul2base_config.yaml#L22).
4. To clarify, I assume with RPE you mean Relative Position Embedding and not Rotary Position Embeddings. We did select [relative position embeddings](https://huggingface.co/Lauler/UL2-nemo-conversion/blob/main/nemo_config/ul2-base-nl36/megatron_model_ul2base_config.yaml#L9). 
5. [We used swiglu](https://huggingface.co/Lauler/UL2-nemo-conversion/blob/main/nemo_config/ul2-base-nl36/megatron_model_ul2base_config.yaml#L24) as this seemed to be supported both in HF (under the name silu, gated-silu activation), and also as an option in the HF -> NeMo conversion script that was available.  

We tried cyclic conversion for an existing UL2 model on HF Hub. We wrote up instructions on replicating our cyclic conversion under the header ["Confirming the conversion script can reverse Nvidia's conversion script"](https://huggingface.co/Lauler/UL2-nemo-conversion#confirming-the-conversion-script-can-reverse-nvidias-conversion-script). However, I only confirmed that all the cyclically converted weights (loaded with T5ForConditionalGeneration) were identical to and matched the weights of the original. That is, I did not try to predict with with the converted model, I just ran [torch.allclose() on all keys in the state dicts](https://huggingface.co/Lauler/UL2-nemo-conversion/blob/main/convert_finnish_ul2_model.py#L25-L35). 

Note that I did not continue pretraining in NeMo with the HF->NeMo converted model. I just did a straight up cyclical conversion HF -> NeMo -> HF to confirm the weight conversion could be reversed.

**Comment by MaximumEntropy** [comment]
Okay, yeah this points to subtle differences in the way the forward passes are implemented between NeMo and HF. IMO the better way to handle normalize_attention_scores is to add the normalization to HF rather than train with the normalization off in NeMo since it may lead to training instabilities. We don't currently have a way to initialize Q,K,V weights with the scaling factor that is used by T5x/PAX/HF T5. 

Another thing you can try is to train in NeMo with `megatron_legacy=True`. This should ensure Q,K,V weights are consistent across NeMo and HF.

**Comment by cpenguf** [comment]
Have you solved the conversion issues?

**Comment by Lauler** [comment]
I think I got close, but gave up because I couldn't pinpoint where my modified forward pass was diverging from Nemo. I got sick of staring at hidden_states and stepping through Nemo library line by line. Here are some insights for anyone that wants to take a crack at it: 

The issue wasn't solely related to the normalization factor. As @MaximumEntropy hinted, the newer versions of Nemo/Megatron perform some strange reshaping of the hidden states that result from multiplying the inputs to an attention layer with the W_q, W_k and W_v matrices. In Nemo the results from those are stored in [`mixed_x_layer`](https://github.com/NVIDIA/NeMo/blob/d81ea6a3f786b46cd204d41d5dc18268763f401c/nemo/collections/nlp/modules/common/megatron/attention.py#L411-L432). 

This tensor starts out fairly harmless as collection of concatenated Q_hidden, K_hidden and V_hidden results. If `hidden_dim` is 768, this means `Q_hidden` occupies 0:768, `K_hidden` 768:1536, etc of the last dimension. 

The problems start when this tensor is reshaped. In Huggingface, they follow the convention of the Transformer paper and ensure a mix of pieces from the hidden states of `Q_h`, `K_h`, `V_h` are sent to different attention heads.  I.e. they send `Q_h[0:64]`, `K_h[0:64]` and `V_h[0:64]` to head number 1, `Q_h[64:128]`, `K_h[64:128]` and `V_h[64:128]` to head number 2, etc. 

Nemo is not as careful to conceptually follow the transformer paper. Nemo rearranges the `mixed_x_layer` to effectively send `Q_h[0:192]` to head nr 1, `Q_h[192:384]` to head nr 2, ..., `K_h[0:192]` to head nr 5, ... `V_h[0:192]` to head nr 9, ..., `V_h[576:768]` to head nr 12. 

So the default multi head attention in Megatron/Nemo technically does not actually end up doing `softmax(QK^T/sqrt(len(head_dim)) V`. Instead it does variants of `(QQ^T)Q`, `(KK^T)K`, `(VV^T)V`. 

There isn't necessarily anything wrong with doing it this way, as labeling the weight matrices Wq, Wk, Wv and keeping track of which matrix produced which hidden state is just a matter of convention. As long as the division is internally consistent, training and inference will work in **your library**. 

Unfortunately the same cannot be said for porting the model to a library that follows a different convention. Even if you port the model weights perfectly to the other library, the forward pass will end up rearranging the hidden states differently and end up multiplying them with the wrong set of weights. 

You can fix this by changing the forward pass of the Huggingface transformers T5 implementation to arrange the matrices the same way as Nemo does. I uploaded  a [fork of transformers which does this](https://github.com/Lauler/transformers/tree/nemo_ul2). Using this I manage to get *almost* the same output from running the first attention block layer in Huggingface as I do running the same layer in Nemo. I fixed the following gotchas:

* Added `norm_factor` for attention scores.
* Rearranged matrices to the exact same shape as in Nemo, because my head was hurting trying to fix it any other way.
* Set `layer_norm_epsilon` to `1e-5` in your Huggingface model's config.json. 

With each fix the numbers kept getting closer to Nemo's output.  I set breakpoints in both model implementations and ran through their attention blocks line by line in parallel. However, I eventually hit a wall and I'm not sure why they start diverging by a very small amount somewhere around where the residual connection is added in the attention block. The difference is very small at first (the first layer), but the divergence accumulates throughout the layers.

If anyone wants to take a crack at it, here are my changes: https://github.com/huggingface/transformers/compare/main...Lauler:transformers:nemo_ul2 . This is for a model trained with `megatron_legacy: false`.

**Comment by TianduoWang** [comment]
Hi Lauler,

Thanks for your brilliant efforts towards converting nemo checkpoints to hf checkpoints. Recently I also tried to convert the model checkpoints and finally made it by following your efforts!

I have gone through your model conversion scripts and test scripts, and found out why you didn't get the exactly the same output tensors from nemo model and hf model.

First, I noticed that you used fp16 precision to train the UL2 model using nemo framework from your nemo config. So when you load the nemo checkpoints for evaluation, the model automatically uses fp16 forward. However, your hf model is loaded in fp32 and does model forward with fp32. This can explain why your nemo model and hf model have small difference even from the first layer. After I also make hf model forward with fp16 precision, the output of the encoder of two models are exactly the same.

Second, I believe there are some implementation errors are made in your modified T5Attention module. In HF T5, both self-attention module and cross-attention module are constructed from the T5Attention module. Your modification is 100% correct for the self-attention module. But it goes wrong when applied to the cross-attention module. This is because, in nemo, when doing the cross-attention, the query weights and key-value weights are two separate parameters in the model. So I think you may need to deal with them separately in cross-attention module in your hf model.

After making these two changes, I got exactly the same outputs from nemo and hf models.

I am not very sure if you are still working on this project, but I am appreciate your very first attempts on this direction and made all your efforts open-sourced. Feel free to let me know if you want to know exactly how I modified your code. I am considering publish my conversion scripts later.

**Comment by Lauler** [comment]
@TianduoWang You are a hero. 

For the time being I've abandoned UL2 with Nemo. But it's always nice to have the option to return to it! I encourage you to publish your conversion script, and please do share if possible! 

I did notice the fp32 vs fp16 issue when I wrote my last comment in this thread. However I was unaware of the cross-attention being different. Nice catch and thanks for sharing the details!

**Comment by viyjy** [comment]
@TianduoWang and @Lauler, thanks for your efforts, very helpful. I am converting a Nemo-CLIP to HF-CLIP and facing the similar issue.
@TianduoWang Would you please share your conversion script? Thanks very much!
